\documentclass{article}
\usepackage{enumitem}
\usepackage{ amssymb }
\usepackage{ textcomp }
\usepackage{longtable}
\usepackage{amsmath,tabu}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{float}
\usepackage[figurename=Supplemental Figure]{caption}
\usepackage{titling}

\setlength{\droptitle}{-5em}  
\topmargin=-0.4in
\evensidemargin=0in
\oddsidemargin=0in
\textwidth=6.5in
\textheight=9in
\headsep=0.25in

\title{Interrogation of human hematopoietic traits at single-cell and single-variant resolution}
\date{Supplemental Information}
\author{(Caleb, Erik, Jacob), Will?, Hilary, Joel, (Martin, Jason, Vijay)}


\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle
\section*{Overview of UK Biobank data}
\begin{enumerate}[label=(\Alph*)]
\item 

<<loadLibraries, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, eval = TRUE, cache = FALSE, results="hide">>=
library(corrplot)
library(BuenColors)
library(dplyr)
library(reshape2)
library(cowplot)
library(data.table)
library(plotly)
@

<<correlationPlots, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Phenotypic and genetic correlations across the 16 traits examined. ", eval = TRUE, cache = TRUE, results="hide">>=

### Phenotypic correlations
phenocors <-read.table("../data/phenotypeCorrelations/raw_phenotypes.txt",header=TRUE,row.names=1)
dissimilarity <- 1 - cor(data.matrix(phenocors))
distance <- as.dist(dissimilarity) 
reorderedtraits=traits=rownames(phenocors)[hclust(distance)$order]

phenocors <- data.matrix(phenocors[reorderedtraits,reorderedtraits])
melt_pheno <- reshape2::melt(phenocors)
melt_pheno$Var1 <- ordered(as.character(melt_pheno$Var1), ordered(traits))
melt_pheno$Var2 <- ordered(as.character(melt_pheno$Var2), ordered(rev((traits))))

p1 <- ggplot(data = melt_pheno, aes(x=Var1, y=Var2, fill=value)) + pretty_plot() + 
  geom_tile() + scale_fill_gradientn(colors = jdb_palette("solar_flare"), limits = c(-1,1)) +
  theme(plot.subtitle = element_text(vjust = 1), 
    plot.caption = element_text(vjust = 1), 
    axis.text.x = element_text(angle = 90)) +labs(x = NULL, y = NULL, fill = "Pearson") +
  theme(legend.position="bottom")  + ggtitle("Phenotype Correlation")

### Constrained intercept genetic correlations
constrained_gencors <- vector(mode = "list", length = length(traits))

# Read unix files into R lists
constrained_gencors <- lapply(traits, function(y) read.table(paste0("../data/phenotypeCorrelations/ldscore/",y,"_UK10K_constrained.gcsummary.txt"),header=TRUE,row.names=NULL))

make.gc.matrix <- function(list_of_gcs,traits){
  allgencors <- bind_rows(list_of_gcs)
  allgencors$rg <- as.numeric(as.character(allgencors$rg))
  
  gc.matrix <- matrix(ncol = 16, nrow = 16, dimnames=list(traits,traits))
  pv.matrix <- matrix(ncol = 16, nrow = 16, dimnames=list(traits,traits))
  
  for (i in traits) {
    for (j in traits) {
      if (i == j){
        gc.matrix[i,j]=1
      } else {
        gc.matrix[i,j] <- allgencors[allgencors$p1 == i & allgencors$p2 == j,'rg']
        if (gc.matrix[i,j]>1) gc.matrix[i,j]=1
        pv.matrix[i,j]<- allgencors[allgencors$p1 == i & allgencors$p2 == j,'p']
      }
    }
  }
  return(list(gc.matrix,pv.matrix))
}

# Plot LD Score correlation
ldscorecor <- make.gc.matrix(constrained_gencors,traits)[[1]]
melt_phenoLD <- reshape2::melt(ldscorecor)
melt_phenoLD$Var1 <- ordered(as.character(melt_phenoLD$Var1), ordered(traits))
melt_phenoLD$Var2 <- ordered(as.character(melt_phenoLD$Var2), ordered(rev((traits))))

p2 <- ggplot(data = melt_phenoLD, aes(x=Var1, y=Var2, fill=value)) + pretty_plot() + 
  geom_tile() + scale_fill_gradientn(colors = jdb_palette("solar_flare"), limits = c(-1,1)) +
  theme(plot.subtitle = element_text(vjust = 1), 
    plot.caption = element_text(vjust = 1), 
    axis.text.x = element_text(angle = 90)) +labs(x = NULL, y = NULL, fill = "LDScore Correlation") +
  theme(legend.position="bottom")  + ggtitle("Genetic Correlation")


### Compare geno and pheno correlations
phenotriangle <- phenocors
phenotriangle[lower.tri(phenotriangle, diag = FALSE)]<- 0
genotriangle <- ldscorecor
genotriangle[lower.tri(genotriangle, diag = FALSE)]<- 0

phenomelt <- melt(phenotriangle); genomelt <- melt(genotriangle); allmelt <- melt(phenocors)

genomelt$traitpairs <- paste(genomelt$Var1,genomelt$Var2)
phenomelt$traitpairs <- paste(phenomelt$Var1,phenomelt$Var2)

colnames(genomelt)[3]<- "GenoCor"
colnames(phenomelt)[3] <- "PhenoCor"
merged <- merge(genomelt,phenomelt,by="traitpairs")
merged[,c("Var1.x","Var2.x")] <- NULL
merged <- merged[merged$GenoCor != 0,]

p3 <- ggplot(merged,aes(x=GenoCor,y=PhenoCor,pairs=traitpairs)) + geom_point(aes(color=Var1.y)) + 
  scale_color_manual(values = UKBB_color_maps) + 
  pretty_plot()+ geom_smooth(method=lm)+
  geom_abline(intercept=0,slope=1) + 
  theme(plot.title = element_text(size=10,hjust = 0.5,face="bold"),
        panel.background = element_rect(fill = "white", colour = "grey50"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_x_continuous(limits = c(-1,1)) + scale_y_continuous(limits = c(-1,1))

 plot_grid(p1, p2, labels = c('A', 'B'))
@


<<pleiotropyImport, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Numbers of pleiotropic variants", eval = TRUE, cache = FALSE, results="hide">>=
 library(ggplot2)
 library(BuenColors)
  
 pleioMatch <- function(file1, file2){
   nnames <- c("chr", "start", "end", "region", "PP", "beta", "SE", "Z")
   # Import Bed and Make GRanges
   hitdf1 <- data.table::fread(as.character(file1))
   names(hitdf1) <- nnames 
   
   hitdf2 <- data.table::fread(as.character(file2))
   names(hitdf2) <- nnames 
   
   hitdf <- merge(hitdf1, hitdf2, by.x = c("chr", "start", "end"), by.y = c("chr", "start", "end"))
   oppo <- sum(sign(hitdf$beta.x) != sign(hitdf$beta.y))
   same <- sum(sign(hitdf$beta.x) == sign(hitdf$beta.y))
   return(c(oppo, same))
 }
 
 files <- list.files("../data/UKBB_BC_PP001/betas_added",  full.names = TRUE)
 pairs <- data.frame(t(combn(files, 2)), stringsAsFactors = FALSE)
 pairs$short1 <- gsub("_PP001_betas.bed", "", gsub("../data/UKBB_BC_PP001/betas_added/", "", pairs[,1]))
 pairs$short2 <- gsub("_PP001_betas.bed", "", gsub("../data/UKBB_BC_PP001/betas_added/", "", pairs[,2]))
 pairs$traits <- paste0(pairs$short1, "_", pairs$short2)
 
 hitsdf <- sapply(1:dim(pairs)[1], function(i){
   pleioMatch(pairs[i,1], pairs[i,2])
 })
 pairsdf <- data.frame(pairs, t(hitsdf))
 colnames(pairsdf) <- c(colnames(pairs), "opposite", "same")
 pairsdf <- pairsdf[,c("short1", "short2", "opposite", "same")]
 
melt_pleio <- reshape2::melt(pairsdf, id.vars = c("short1", "short2"))
melt_pleio$short1 <- ordered(as.character(melt_pleio$short1), ordered(traits))
melt_pleio$short2 <- ordered(as.character(melt_pleio$short2), ordered(rev((traits))))

mdf <- rbind(melt_pleio, setNames(melt_pleio[,c(2,1,3,4)], c("short1", "short2", "variable", "value")))
mdf$value[is.na(mdf$value)] <- 0
@

<<pleiotropyPlots, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Numbers of pleiotropic variants", eval = TRUE, cache = FALSE, results="hide">>=
pleioplot1 <- ggplot(data = mdf[mdf$variable == "same", ], aes(x=short1, y=short2, fill=value)) + pretty_plot() + 
  geom_tile() + scale_fill_gradientn(colors = jdb_palette("solar_rojos")) +
  theme(plot.subtitle = element_text(vjust = 1), legend.text=element_text(size= 7),
    plot.caption = element_text(vjust = 1), 
    axis.text.x = element_text(angle = 90)) +labs(x = NULL, y = NULL, fill = "# of Variants") +
  theme(legend.position="bottom")  + ggtitle("Same sign")
 

pleioplot2 <- ggplot(data = mdf[mdf$variable == "opposite", ], aes(x=short1, y=short2, fill=value)) + pretty_plot() + 
  geom_tile() + scale_fill_gradientn(colors = jdb_palette("solar_blues")) +
  theme(plot.subtitle = element_text(vjust = 1),  legend.text=element_text(size= 7),
    plot.caption = element_text(vjust = 1), 
    axis.text.x = element_text(angle = 90)) +labs(x = NULL, y = NULL, fill = " # of Variants") +
  theme(legend.position="bottom")  + ggtitle("Opposite sign")
 plot_grid(pleioplot1, pleioplot2, labels = c('A', 'B'))
@

<<heritabilityPlots, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Heritability estimates from LD Score Regression across 16 hematopoetic traits. The estimates of the narrow-sense SNP heritabilities are plotted in gray with their corresponding standard erors. Heritability estimates for all variants with fine-mapped posterior probability > 0.001 are plotted in blue for each trait, and the proportions of total narrow-sense heritability captured by these fine-mapped variants (blue bar / gray bar) are indicated by the numbered labels.", eval = TRUE, cache = FALSE, results="hide">>=

library(scales)
# Narrow-Sense Heritability Estimated Obtained from LDSC
h2 <-fread("../data/LDSC_heritability/trait_h2.txt")
colnames(h2) <- c("trait","h_obs","h_obs_se")

# Read in enrichments for FM PP001 variants
reorderedtraits=traits=c("BASO_COUNT","EO_COUNT","HCT","HGB","LYMPH_COUNT", "MCH", "MCHC", "MCV", "MEAN_RETIC_VOL","MONO_COUNT", "MPV", "NEUTRO_COUNT", "PLT_COUNT", "RBC_COUNT","RETIC_COUNT","WBC_COUNT")

ctHeme <- vector("list",length=16)
ct_enrichment <- vector("list",length=length(ctHeme))
celltypes <- paste0(traits,"_PP001L2")

FM_h2 <- NULL
FM_enrichment <- NULL
dir="../data/LDSC_heritability/FM_PP001/"
for (i in 1:length(ctHeme)){
  ctHeme[[i]]<-fread(paste0(dir,traits[i],".",celltypes[i],".results"))
  FM_h2[i] <- ctHeme[[i]]$Prop._h2[1]
  FM_enrichment[i] <- ctHeme[[i]]$Enrichment[1]
}

h2$fm_herit <- FM_h2*h2$h_obs
h2$fm_h2 <- FM_h2

legendtitle <- "fine-mapped heritability"
ggplot(data=h2, aes(x=trait,y=fm_herit)) +
  geom_bar(aes(y=h_obs), fill="light gray",stat="identity",position = position_stack(reverse = TRUE)) +
  geom_bar(aes(y=fm_herit,fill=legendtitle),stat="identity") +
  geom_text(aes(label=percent(FM_h2)), hjust=-0.25, color="black", size=3.25) +
  scale_fill_manual(values = c(jdb_palette("GrandBudapest2")[c(4)],jdb_palette("Zissou")[1:5])) +
  ggtitle("Finemapped Trait Heritabilities") + theme_bw() + 
  coord_flip() + 
  labs(fill="") + 
  scale_y_continuous(expand = c(0.01, 0.0)) +
  theme(plot.subtitle = element_text(vjust = 1),
        panel.background = element_rect(fill = "white", colour = "grey50"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        legend.position = c(.98, 0.99),
        legend.justification = c("right", "top"),
        legend.box.just = "right",
        legend.margin = margin(1, 1, 1, 1),
        legend.direction = "vertical",
        legend.key.size = unit(0.15, "in"))  +
  geom_errorbar(aes(ymin=h_obs-h_obs_se, ymax=h_obs+h_obs_se),width=.2,position=position_dodge(.9),color="black")

@

<<ccnd3Piechart, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Haplotypes frequencies of the CCND3 variants rs2408955 and rs112233623 in the UKBB Population. Haplotype labels are formatted as rs2408955/rs112233623.", eval = TRUE, cache = FALSE, results="hide">>=

### Moderating alleles
trait <- "RBC_COUNT"
unit <- "10^6 cells/uL"
rs_alleles <- readRDS("../data/examples/CCND3.haplotypes.rds")

# Only look at the imputed genotypes with 0, 1, or 2 for boxplots
rs_alleles_onlyabsolutes <- as.data.frame(rs_alleles[rs_alleles$alleleA.rs112233623 == 0 |
                                                       rs_alleles$alleleA.rs112233623 == 1 | 
                                                       rs_alleles$alleleA.rs112233623 == 2,])

# Write out genotypes for rs1
allelecol1 <- "alleleA.rs2408955"
rs_alleles_onlyabsolutes$rs1_A <- "AA"
rs_alleles_onlyabsolutes[grep(0,rs_alleles_onlyabsolutes[,allelecol1]),"rs1_A"] <- "AA"
rs_alleles_onlyabsolutes[grep(1,rs_alleles_onlyabsolutes[,allelecol1]),"rs1_A"] <- "AG"
rs_alleles_onlyabsolutes[grep(2,rs_alleles_onlyabsolutes[,allelecol1]),"rs1_A"] <- "GG"
rs_alleles_onlyabsolutes$rs1_A <- factor(rs_alleles_onlyabsolutes$rs1_A, levels = c("GG","AG","AA"))

# Write out genotypes for rs2
allelecol2 <- "alleleA.rs112233623"
rs_alleles_onlyabsolutes$rs2_A <- "TT"
rs_alleles_onlyabsolutes[grep(0,rs_alleles_onlyabsolutes[,allelecol2]),"rs2_A"] <- "TT"
rs_alleles_onlyabsolutes[grep(1,rs_alleles_onlyabsolutes[,allelecol2]),"rs2_A"] <- "CT"
rs_alleles_onlyabsolutes[grep(2,rs_alleles_onlyabsolutes[,allelecol2]),"rs2_A"] <- "CC"
rs_alleles_onlyabsolutes$rs2_A <- factor(rs_alleles_onlyabsolutes$rs2_A, levels = c("CC","CT","TT"))

# Make column for all haplotype combinations
#rs_alleles_onlyabsolutes$con_rs1 <- ifelse(rs_alleles_onlyabsolutes$rs1_A =="GG" & rs_alleles_onlyabsolutes$rs2_A =="CT", "GG/CT", "")
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="AA" & rs_alleles_onlyabsolutes$rs2_A=="CT","con_rs1"] <- "AA/CT"
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="AG" & rs_alleles_onlyabsolutes$rs2_A=="CT","con_rs1"] <- "AG/CT"
#rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="GG" & rs_alleles_onlyabsolutes$rs2_A=="CT","con_rs1"] <- "GG/CT"
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="AA" & rs_alleles_onlyabsolutes$rs2_A=="CC","con_rs1"] <- "AA/CC"
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="AG" & rs_alleles_onlyabsolutes$rs2_A=="CC","con_rs1"] <- "AG/CC"
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="GG" & rs_alleles_onlyabsolutes$rs2_A=="CC","con_rs1"] <- "GG/CC"
rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes$rs1_A=="AA" & rs_alleles_onlyabsolutes$rs2_A=="TT","con_rs1"] <- "AA/TT"
rs_alleles_onlyabsolutes$con_rs1 <- factor(rs_alleles_onlyabsolutes$con_rs1, 
                                           levels = c("GG/CC","AG/CC","AA/CC",
                                                      "AG/CT","AA/CT","AA/TT"))

# Pie chart of allele frequencies
proportions <- data.frame(table(rs_alleles_onlyabsolutes$con_rs1))
colnames(proportions) <- c("Genotype","Freq")
proportions$Percent <- proportions$Freq / sum(proportions$Freq)

m <- list(
  l = 50,
  r = 50,
  b = 70,
  t = 25,
  pad = 1
)
plot_ly(proportions, labels = ~Genotype, values = ~Freq, type = 'pie',textposition = 'outside',textinfo = 'label+percent',width = 600, height = 700) %>%
  layout(xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE), autosize = F,  margin = m)
# tmpFile <- tempfile(fileext = ".pdf")
# export(p, file = "/Users/erikbao/Dropbox (MIT)/HMS/Sankaran Lab/ATACSeq_GWAS/Examples/CCND3/rs9349205_rs112233623.conditional_piechart.pdf")

@

<<ccnd3Boxplots, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Moderating effects of CCND3 haplotypes on red blood cell count.", eval = TRUE, cache = FALSE, results="hide">>=

# The baseline is the homozygous ref/ref haplotypes.
baseline <- as.data.frame(rs_alleles_onlyabsolutes[rs_alleles_onlyabsolutes[,"con_rs1"] %in% "GG/CC",trait])[,1] %>% median()

# Calculate median phenotype value for each haplotype. 
meds <- rs_alleles_onlyabsolutes %>% group_by(con_rs1) %>% summarise(median = median(RBC_COUNT, na.rm = TRUE))
meds <- meds[1:5,]

# Set y-axis bounds
limits <- c(as.numeric(quantile(rs_alleles_onlyabsolutes[,trait],0.001)),
            as.numeric(quantile(rs_alleles_onlyabsolutes[,trait],0.999)))

ggplot(subset(rs_alleles_onlyabsolutes,rs2_A=="CC" | rs2_A=="CT"), 
              aes_string(x="con_rs1", trait)) + 
  geom_boxplot() + pretty_plot() +
  geom_boxplot(outlier.colour=NA) + 
  labs(x="Haplotype (rs2408955/rs112233623)",
       y=paste0("RBC Count (",unit,")")) +
  ggtitle("rs2408955/rs112233623 Haplotype Effects") +
  geom_hline(yintercept = baseline,linetype="dashed") +
  geom_text(data = meds, aes(x = con_rs1, y = baseline, label = paste(round(median,3))), size = 3, vjust = -1.8) +
  coord_cartesian(ylim = limits)+
    theme(plot.subtitle = element_text(vjust = 1))

@

<<goshifterLocalZscore, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Local z-scores for enrichment of hematopoietic nucleosome-depleted regions in the set of fine-mapped variants with posterior probability $> 0.10$.", eval = TRUE, cache = FALSE, results="hide">>=

library(rtracklayer)
library(preprocessCore)
library(ggjoy)
library(regioneR)
library(stringr)

#' Read in count matrix 
#+ cache = FALSE, message = FALSE, warning = FALSE, echo = FALSE, eval = TRUE
counts.df <- read.table("../data/bulk/ATAC/26August2017_EJCsamples_allReads_250bp.counts.txt",header=T)
names(counts.df) <- c("B","CD4","CD8","CLP","CMP","Ery","GMP-A","GMP-B","GMP-C","HSC","LMPP","mDC","Mega","MEP","mono","MPP","NK","pDC")
# Remove "weak" peaks that aren't in top n% for at least one cell type
n=0.8
counts.df[1:dim(counts.df)[1],1:dim(counts.df)[2]] <- normalize.quantiles(as.matrix(counts.df))
keep <- apply(counts.df,1,max) > mean(apply(counts.df,2,function(x) {quantile(x,n)}))

#' Read in consensus bed file of peaks and add count metadata
#+ cache = FALSE, message = FALSE, warning = FALSE, echo = FALSE, eval = TRUE
peaks.gr <- import("../data/bulk/ATAC/26August2017_EJCsamples_allReads_250bp.bed",format="bed")
values(peaks.gr) <- counts.df
peaks.gr <- peaks.gr[keep,]

# Read in Finemap variants
CS.df <- read.table("../data/Finemap/UKBB_BC_v3.bed")
names(CS.df) <- c("seqnames","end","start","annotation","PP")
CS.df[,c("trait","var","region")] <- str_split_fixed(CS.df$annotation, "-", 3)
CS.df <- CS.df %>% dplyr::select(-annotation)
CS.df$seqnames <- paste0("chr",CS.df$seqnames)
CS.gr <- makeGRangesFromDataFrame(CS.df, keep.extra.columns=TRUE, ignore.strand=TRUE, starts.in.df.are.0based=TRUE)

perm <- NULL
bins = c(0.001,0.01,0.05,0.1,0.25,0.75,1)

randomizeLocalRegions2 <- function(A, ...) {
  A <- toGRanges(A)
  rand <- ceiling(runif(1,-1500000,1500000))
  B <- GenomicRanges::shift(A,rand)
  return(B)
}

for (j in seq(4,4,1)) {
  #CS.PP.gr <- subset(CS.gr,PP > bins[j] & PP <= bins[j+1]) 
  CS.PP.gr <- subset(CS.gr,PP > bins[j]) 
  perm[[j]] <- permTest(A=CS.PP.gr, B=peaks.gr, ntimes=1000, alternative="auto", randomize.function=randomizeLocalRegions2, evaluate.function=numOverlaps, force.parallel=FALSE, mc.set.seed=TRUE, mc.cores=2)
}

local <- localZScore(CS.PP.gr, peaks.gr, pt=perm[[4]], window=1500000,step=1000, mc.set.seed=TRUE, mc.cores=2)
plot(local, main="Local Shifted Enrichments for Fine-mapped PP>0.10 Variants")

@

<<fmSecondaryEnrichments, message = FALSE, warning = FALSE, echo = FALSE, fig.height=5, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Local annotation enrichments for fine-mapped variants, excluding all fine-map variants with $R^2>0.80$ to the sentinel variant of any region.", eval = TRUE, cache = FALSE, results="hide">>=

# Exclude LD80 variants from CS
LD80 <- readRDS("../data/Finemap/LD80.rds")
LD80.gr <- GRanges()
for (i in 1:16){
  temp <- as.data.frame(LD80[[i]])[,c("chromosome","sentinel","nonsentinel")]
  temp <- melt(temp,id.vars=c("chromosome"))
  temp$chromosome <- paste0("chr",temp$chromosome)
  temp$end <- temp$value
  temp.gr <- makeGRangesFromDataFrame(temp, keep.extra.columns=F,
                                   ignore.strand=TRUE, 
                                   starts.in.df.are.0based=F,
                                   start.field="value",
                                   end.field="end")
  
  LD80.gr <- mergeRegions(LD80.gr, temp.gr)
}  

#' Read in other annotations
#+ cache = FALSE, message = FALSE, warning = FALSE, echo = FALSE, eval = TRUE
coding.gr <- import("../data/annotations/Coding_UCSC.bed",format="bed")
intron.gr <- import("../data/annotations/Intron_UCSC.bed",format="bed")
promoter.gr <- import("../data/annotations/Promoter_UCSC.fixed.bed",format="bed")
UTR.gr <- import("../data/annotations/UTR_3_UCSC.bed",format="bed")

randomizeLocalRegions2 <- function(A, ...) {
  A <- toGRanges(A)
  rand <- ceiling(runif(1,-1500000,1500000))
  B <- GenomicRanges::shift(A,rand)
  return(B)
}

OR <- NULL
perm <- NULL
trait <- unique(CS.gr@elementMetadata$trait)
gr.list <- list(peaks.gr,coding.gr,promoter.gr,intron.gr,UTR.gr)
bins = c(0.001,0.01,0.05,0.1,0.25,0.75,1)
for (i in seq(1,5,1)) {
  for (j in seq(1,6,1)) {
    CS.PP.gr <- subset(CS.gr,PP > bins[j] & PP <= bins[j+1]) 
    CS.PP.gr <- subtractRegions(CS.PP.gr,LD80.gr)
    perm[[paste0(i,j)]] <- permTest(A=CS.PP.gr, B=gr.list[[i]], ntimes=100, alternative="auto", randomize.function=randomizeLocalRegions2, evaluate.function=numOverlaps, force.parallel=FALSE, mc.set.seed=FALSE, mc.cores=4)
    OR <- rbind(OR,c(i,j,perm[[paste0(i,j)]]$numOverlaps$observed / length(CS.PP.gr), mean(perm[[paste0(i,j)]]$numOverlaps$permuted) / length(CS.PP.gr), perm[[paste0(i,j)]]$numOverlaps$zscore))
  print(c(i,j))
  }
}

OR.df <- as.data.frame(OR)
names(OR.df) <- c("i","j","obs","perm")
OR.df$i <- as.factor(OR.df$i)
OR.df$j <- as.factor(OR.df$j)
OR.df$OR <- OR.df$obs / OR.df$perm
OR.df <- OR.df %>% dplyr::select(i,j,OR)

ggplot(OR.df,aes(y=OR,x=i,group=j)) + 
  geom_bar(stat="identity",aes(fill=j),position = position_dodge(0.9)) +
  geom_hline(yintercept=1, colour="grey3", linetype =3) +
  theme_bw() +
  pretty_plot() +
  scale_y_continuous(expand = c(0.01, 0.0)) +
  scale_x_discrete(labels=c("heme NDRs","coding","promoter","intron","UTR"))+
  scale_fill_manual(values = c(jdb_palette("GrandBudapest2")[c(4)],jdb_palette("Zissou")[1:5]), 
                    name = "top posterior",
                    labels=c("(0.001,0.01]", "(0.01,0.05]", "(0.05,0.1]","(0.1,0.25]","(0.25,0.75]","(0.75,1]")) +
  labs(y="odds ratio") +
  theme(axis.title.x=element_blank(),
        axis.ticks.x =element_blank())

@

\item
\begin{figure}[H]
\centering
\includegraphics[width=\linewidth]{staticFigures/UKBB_PP50_bulk.pdf}
\caption{Side-by-side heatmaps showing overlap of hematopoietic nucleosome-depleted regions by cell type with fine-mapped variants ($PP>0.50$) by trait. The two heatmaps share a common y-axis of specific variants.}
\end{figure} 

<<barplots, message = FALSE, warning = FALSE, echo = FALSE, fig.height=9, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Hematopoetic cell type enrichments for Mean Retic Volume and Monocyte count using various methods.", eval = TRUE, cache = FALSE, results="hide">>=
df3 <- read.table("../data/bulk/GWAS-Bulk/compare3.tsv", header = TRUE)
df3$LDScore <- -log10(df3$ldscore_pvalue)
df3$`g-chromVAR` <- -log10(pnorm(df3$weighted_Zscore, lower.tail = FALSE))
df3$chromVAR <- -log10(pnorm(df3$chromVAR_Zscore, lower.tail = FALSE))
plotdfEnrich2 <- reshape2::melt(df3[df3$Trait %in% c("MEAN_RETIC_VOL", "MONO_COUNT"),c(1,2,6,8,7)],
                                id.vars = c("Celltype", "Trait"))

pEnrich1 <- ggplot(plotdfEnrich2[plotdfEnrich2$Trait == "MONO_COUNT", ],
                   aes(x = Celltype, y = value)) +
  geom_bar(width = 1, aes(fill = Celltype), colour="black",
           stat = "identity", position = position_dodge(width=1)) +
  pretty_plot() + labs(x = "", y = "Monocyte Count Enrichment (-log10 p)", fill = "") +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) +
   scale_fill_manual(values = ejc_color_maps) + facet_grid(variable ~ ., scales = "free_y") +
  theme(legend.position="none") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

pEnrich2 <- ggplot(plotdfEnrich2[plotdfEnrich2$Trait == "MEAN_RETIC_VOL", ],
                   aes(x = Celltype, y = value)) +
  geom_bar(width = 1, aes(fill = Celltype), colour="black",
           stat = "identity", position = position_dodge(width=1)) +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) +
  pretty_plot() + labs(x = "", y = "Mean Retic Volume Enrichment (-log10 p)", fill = "") +
   scale_fill_manual(values = ejc_color_maps) + facet_grid(variable ~ ., scales = "free_y") +
  theme(legend.position="bottom") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

plot_grid(pEnrich1, pEnrich2, labels = c('A', 'B'), nrow = 2, rel_heights = c(1,1.3))
@

<<allGchromvar, message = FALSE, warning = FALSE, echo = FALSE, fig.height=8, fig.align='center', fig.pos='H', fig.width=8, fig.cap="All enrichments from g-chromVAR. The horizontal line shows a Bonferonni multiple testing adjusted threshold for statistical significance of enrichment.", eval = TRUE, cache = FALSE, results="hide">>=

# Plot of all enrichments
dfg <- read.table("../data/bulk/GWAS-Bulk/compare3.tsv", header = TRUE)
dfg$`g-chromVAR` <- -log10(pnorm(dfg$weighted_Zscore, lower.tail = FALSE))
plotdfEnrichg <- reshape2::melt(dfg[,c(1,2,6)],
                                id.vars = c("Celltype", "Trait"))
ggplot(plotdfEnrichg, aes(x = Celltype, y = value)) +
  geom_bar(width = 1, aes(fill = Celltype), colour="black",
           stat = "identity", position = position_dodge(width=1)) +
  pretty_plot() + labs(x = "", y = "g-chromVAR Enrichment (-log10 p)", fill = "") +
   scale_fill_manual(values = ejc_color_maps) + facet_wrap(~Trait, scales = "free_y") +
  theme(legend.position="bottom") +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
@

<<allchromVAR, message = FALSE, warning = FALSE, echo = FALSE, fig.height=8, fig.align='center', fig.pos='H', fig.width=8, fig.cap="All enrichments from chromVAR. The horizontal line shows a Bonferonni multiple testing adjusted threshold for statistical significance of enrichment.", eval = TRUE, cache = FALSE, results="hide">>=

# Plot of all enrichments
dfg <- read.table("../data/bulk/GWAS-Bulk/compare3.tsv", header = TRUE)
dfg$chromVAR <- -log10(pnorm(dfg$chromVAR_Zscore, lower.tail = FALSE))
plotdfEnrichg <- reshape2::melt(dfg[,c(1,2,6)],
                                id.vars = c("Celltype", "Trait"))
ggplot(plotdfEnrichg, aes(x = Celltype, y = value)) +
  geom_bar(width = 1, aes(fill = Celltype), colour="black",
           stat = "identity", position = position_dodge(width=1)) +
  pretty_plot() + labs(x = "", y = "chromVAR Enrichment (-log10 p)", fill = "") +
   scale_fill_manual(values = ejc_color_maps) + facet_wrap(~Trait, scales = "free_y") +
  theme(legend.position="bottom") +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
@

<<allLDscore, message = FALSE, warning = FALSE, echo = FALSE, fig.height=8, fig.align='center', fig.pos='H', fig.width=8, fig.cap="All enrichments from LD Score. The horizontal line shows a Bonferonni multiple testing adjusted threshold for statistical significance of enrichment.", eval = TRUE, cache = FALSE, results="hide">>=

# Plot of all enrichments
dfldscore <- read.table("../data/bulk/GWAS-Bulk/compare3.tsv", header = TRUE)
dfldscore$ldscoreP<- -log10(dfldscore$ldscore_pvalue)
plotdfEnrichLDscore <- reshape2::melt(dfldscore[,c(1,2,6)], id.vars = c("Celltype", "Trait"))
ggplot(plotdfEnrichLDscore, aes(x = Celltype, y = value)) +
  geom_bar(width = 1, aes(fill = Celltype), colour="black",
           stat = "identity", position = position_dodge(width=1)) +
  pretty_plot() + labs(x = "", y = "LD Score Enrichment Enrichment (-log10 p)", fill = "") +
   scale_fill_manual(values = ejc_color_maps) + facet_wrap(~Trait, scales = "free_y") +
  theme(legend.position="bottom") +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
@

<<varyingPP, message = FALSE, warning = FALSE, echo = FALSE, fig.height=8, fig.align='center', fig.pos='H', fig.width=8, fig.cap="Cell type - trait enrichments for g-chromVAR across different finemap variants posterior probability cutoffs. The horizontal line shows a Bonferonni multiple testing adjusted threshold for statistical significance of enrichment.", eval = TRUE, cache = FALSE, results="hide">>=
zscoreWeighted <- readRDS("../data/supplement_rds/varyingPPcutoff.rds")
zscoreWeighted$Var2 <- gsub("_PP001", "", zscoreWeighted$Var2)
ggplot(zscoreWeighted, aes(x = PP, y = -log10(pnorm(value,lower.tail = FALSE)))) +
  geom_line(aes(color = Var1)) +
  pretty_plot() + labs(x = "Posterior Probability Threshold", y = "g-chromVAR Enrichment (-log10 p)", color = "") +
   scale_color_manual(values = ejc_color_maps) + facet_wrap(~Var2, scales = "free_y") +
  theme(legend.position="bottom") +
  geom_hline(yintercept = -log10(0.05 / (16*17)), linetype = 2) 

@
\item

\begin{figure}[H]
\centering
\includegraphics[width=\linewidth]{staticFigures/LDtoSentinel_PPcomparison.pdf}
\caption{Sample embedding of figure in document.}
\end{figure} 

\end{enumerate}

\end{document}